File: core/src/ru/deadsoftware/cavecraft/game/objects/Drop.java
package ru.deadsoftware.cavecraft.game.objects;

import com.badlogic.gdx.math.Intersector;
import com.badlogic.gdx.math.Rectangle;
import com.badlogic.gdx.math.Vector2;
import ru.deadsoftware.cavecraft.game.GameProc;

import java.io.Serializable;

public class Drop implements Serializable {
    private int id;
    public boolean pickedUp = false;
    public Vector2 move, pos;

    public Drop(float x, float y, int id) {
        this.id = id;
        pos = new Vector2(x, y);
        move = new Vector2(0, -1);
    }

    public int closeToPlayer(GameProc gp) {
        boolean c1 = Intersector.overlaps(new Rectangle(gp.player.pos.x - 16, gp.player.pos.y - 16, gp.player.texWidth + 32, gp.player.height + 32), getRect());
        boolean c2 = Intersector.overlaps(new Rectangle((gp.player.pos.x + gp.world.getWidth() * 16) - 16, gp.player.pos.y - 16, gp.player.texWidth + 32, gp.player.height + 32), getRect());
        boolean c3 = Intersector.overlaps(new Rectangle((gp.player.pos.x - gp.world.getWidth() * 16) - 16, gp.player.pos.y - 16, gp.player.texWidth + 32, gp.player.height + 32), getRect());
        if (c1) return 1;
        if (c2) return 2;
        if (c3) return 3;
        return 0;
    }

    public void moveToPlayer(GameProc gp) {
        int ctp = closeToPlayer(gp);
        if (ctp > 0) {
            float px = gp.player.pos.x;
            float py = gp.player.pos.y;
            switch (ctp) {
                case 2:
                    px += gp.world.getWidth() * 16;
                    break;
                case 3:
                    px -= gp.world.getWidth() * 16;
                    break;
            }
            float dx = 0, dy = 0;
            if (px + gp.player.texWidth < pos.x + 4) dx = -.5f;
            else if (px > pos.x + 4) dx = .5f;
            if (py + gp.player.height < pos.y + 4) dy = -.5f;
            else if (py > pos.y + 4) dy = .5f;
            move.add(dx, dy);
            if (move.x > 2) move.x = 1;
            if (move.x < -2) move.x = -1;
            if (move.y > 2) move.y = 1;
            if (move.y < -2) move.y = -1;
        }
    }

    public void pickUpDrop(Player pl) {
        for (int i = 0; i < pl.inv.length; i++) {
            if (pl.inv[i] == 0 || pl.inv[i] == id) {
                pl.inv[i] = id;
                pickedUp = true;
                break;
            }
        }
    }

    public int getId() {
        return id;
    }

    public Rectangle getRect() {
        return new Rectangle(pos.x, pos.y, 8, 8);
    }

}

File: core/src/ru/deadsoftware/cavedroid/game/objects/drop/Drop.kt
package ru.deadsoftware.cavedroid.game.objects.drop

import com.badlogic.gdx.math.Intersector
import com.badlogic.gdx.math.Rectangle
import com.badlogic.gdx.math.Vector2
import ru.deadsoftware.cavedroid.game.GameItemsHolder
import ru.deadsoftware.cavedroid.game.model.dto.SaveDataDto
import ru.deadsoftware.cavedroid.game.model.item.Item
import ru.deadsoftware.cavedroid.misc.Saveable

class Drop @JvmOverloads constructor(
    x: Float,
    y: Float,
    _item: Item,
    _amount: Int = 1,
) : Rectangle(x, y, DROP_SIZE, DROP_SIZE), Saveable {

    var amount: Int = _amount
        private set

    val itemKey = _item.params.key
    val velocity = getInitialVelocity()
    var pickedUp = false

    @Suppress("UNNECESSARY_LATEINIT")
    @Transient
    lateinit var item: Item
        private set

    init {
        item = _item
    }

    fun initItem(gameItemsHolder: GameItemsHolder) {
        if (this::item.isInitialized) {
            return
        }

        item = gameItemsHolder.getItem(itemKey)
    }

    fun canMagnetTo(rectangle: Rectangle): Boolean {
        val magnetArea = getMagnetArea()
        return Intersector.overlaps(magnetArea, rectangle)
    }

    @JvmOverloads
    fun subtract(count: Int = 1) {
        if (count < 0) {
            throw IllegalArgumentException("Can't subtract negative amount")
        }

        amount -= count
    }

    private fun getMagnetArea(): Rectangle {
        return Rectangle(
            /* x = */ x - MAGNET_DISTANCE,
            /* y = */ y - MAGNET_DISTANCE,
            /* width = */ width + MAGNET_DISTANCE * 2,
            /* height = */ height + MAGNET_DISTANCE * 2,
        )
    }

    override fun getSaveData(): SaveDataDto.DropSaveData {
        return SaveDataDto.DropSaveData(
            version = SAVE_DATA_VERSION,
            x = x,
            y = y,
            width = width,
            height = height,
            velocityX = velocity.x,
            velocityY = velocity.y,
            itemKey = itemKey,
            amount = amount,
            pickedUp = pickedUp
        )
    }

    companion object {
        private const val SAVE_DATA_VERSION = 1

        private const val MAGNET_DISTANCE = 8f

        const val MAGNET_VELOCITY = 256f
        const val DROP_SIZE = 8f

        private fun getInitialVelocity(): Vector2 = Vector2(0f, -1f)

        fun fromSaveData(saveData: SaveDataDto.DropSaveData, gameItemsHolder: GameItemsHolder): Drop {
            saveData.verifyVersion(SAVE_DATA_VERSION)

            return Drop(
                x = saveData.x,
                y = saveData.y,
                _item = gameItemsHolder.getItem(saveData.itemKey),
                _amount = saveData.amount,
            ).apply {
                velocity.x = saveData.velocityX
                velocity.y = saveData.velocityY
                pickedUp = saveData.pickedUp
            }
        }
    }
}
